Plugins/Events/Events/SpellEvents.cpp

```diff
@@ -0,0 +1,61 @@
+#include "Events/SpellEvents.hpp"
+#include "API/CNWSCreature.hpp"
+#include "API/Functions.hpp"
+#include "Events.hpp"
+#include "Helpers.hpp"
+
+namespace Events {
+
+using namespace NWNXLib;
+
+SpellEvents::SpellEvents(ViewPtr<Services::HooksProxy> hooker)
+{
+    hooker->RequestSharedHook<
+        NWNXLib::API::Functions::CNWSObject__SpellCastAndImpact, 
+        int32_t, 
+        NWNXLib::API::CNWSObject*, 
+        uint32_t,
+        NWNXLib::API::Vector,
+        uint32_t,
+        int8_t,
+        uint32_t,
+        bool,
+        bool,
+        int8_t,
+        bool>(SpellEvents::CastSpellHook);
+}
+
+void SpellEvents::CastSpellHook
+(
+    Services::Hooks::CallType type, 
+    NWNXLib::API::CNWSObject* thisPtr, 
+    uint32_t spellID, 
+    NWNXLib::API::Vector targetPosition,
+    NWNXLib::API::Types::ObjectID oidTarget, 
+    int8_t multiClass, 
+    NWNXLib::API::Types::ObjectID oidItem,
+    bool spellCountered,
+    bool counteringSpell,
+    int8_t projectilePathType,
+    bool isInstantSpell
+)
+{
+    const bool before = type == Services::Hooks::CallType::BEFORE_ORIGINAL;
+    Events::PushEventData("SPELL_ID", std::to_string(spellID));
+
+    Events::PushEventData("TARGET_POSITION_X", std::to_string(targetPosition.x));
+    Events::PushEventData("TARGET_POSITION_Y", std::to_string(targetPosition.y));
+    Events::PushEventData("TARGET_POSITION_Z", std::to_string(targetPosition.z));
+
+    Events::PushEventData("TARGET_OBJECT_ID", Helpers::ObjectIDToString(oidTarget));
+    Events::PushEventData("MULTI_CLASS", std::to_string(multiClass));
+    Events::PushEventData("ITEM_OBJECT_ID", Helpers::ObjectIDToString(oidItem));
+    Events::PushEventData("SPELL_COUNTERED", std::to_string(spellCountered));
```

Comment:
By the standard, there isn't an overload of std::to_string for bool. It might be a non-standard extension (in which case, maybe double check that it prints 0/1 rather than "false" and "true"?) but more likely it is implicitly converting bool to an int which is why it works.

I just wanted to call it out here because I can totally see some platform making std::to_string(bool) which outputs the text representation of a bool instead. It might be worth explicitly casting the bool to an int, or doing something like:

`Events::PushEventData("SPELL_COUNTERED", spellCountered ? "1" : "0");`